service/history/timer_queue_task_executor_base.go

```
@@ -240,8 +241,51 @@ func (t *timerQueueTaskExecutorBase) executeSingleStateMachineTimer(
 	return nil
 }
 
-// executeStateMachineTimers gets the state machine timers, processed the expired timers,
-// and return a slice of unprocessed timers.
+// executeChasmPureTimers walks a CHASM tree for expired pure task timers,
+// executes them, and returns a count of timers processed.
+func (t *timerQueueTaskExecutorBase) executeChasmPureTimers(
+	ctx context.Context,
+	workflowContext historyi.WorkflowContext,
+	ms historyi.MutableState,
+	task *tasks.ChasmTaskPure,
+	execute func(node *chasm.Node, task any) error,
+) (processedCount int, err error) {
+	// Because CHASM timers can target closed workflows, we need to specifically
+	// exclude zombie workflows, instead of merely checking that the workflow is
+	// running.
+	if ms.GetExecutionState().State == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
+		return 0, consts.ErrWorkflowZombie
+	}
+
+	tree := ms.ChasmTree()
+	if tree == nil {
+		return 0, consts.ErrStaleReference
```

nit: shall we return an internal error here? I think this is not expected?